<a href="https://colab.research.google.com/github/Theeyecode/Housing-Stress-Canada/blob/eda/descriptive_stat.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exploratory Data Analysis for Canada Housing Survery Data 2022

In [12]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


url = "https://drive.google.com/uc?id=11Y8p_9-CYw0tpGPFu-jlgzxzWOOVS43F"

In [2]:


df = pd.read_csv(url)
df.head()

PUMFID  EHA_10  EHA_25  FP_05  DWI_05A  DWI_05B  DWI_05C  DWI_05D  NEI_05A  \
0   63501       4       2      1        2        2        2        2        4   
1   63502       3       2      2        2        2        2        2        4   
2   63503       3       2      1        2        2        2        2        4   
3   63504       3       2      1        2        1        2        2        3   
4   63505       4       2      1        2        2        2        2        4   

   NEI_05B  ...  PSTIR_GR  PVISMIN  PWSA_D15  P2DCT_20  P2DCT_25  PATT_05  \
0        4  ...         1        9     999.6       996         6        1   
1        4  ...         1        2     999.6       996         6        2   
2        4  ...         1        2     999.6       996         6        2   
3        4  ...         1        1     999.6       996         6        2   
4        4  ...         1        2     999.6         2         1        1   

   PATT_10  PATT_15A  PATT_15B     VERDATE  
0        1         6         6  11/08/2025  
1        1         6         6  11/08/2025  
2        1         6         6  11/08/2025  
3        1         6         6  11/08/2025  
4        2         6         3  11/08/2025  

[5 rows x 103 columns]

Shape of the Raw data


In [3]:
df.shape

(38657, 103)

In [4]:
# df.dtypes
df.dtypes.value_counts()


int64      98
float64     4
object      1
Name: count, dtype: int64

In [5]:
# Identify non-numeric columns (typically dates or text fields)
df.select_dtypes(include="object").columns.tolist()

['VERDATE']

In [6]:
# Convert verification date to datetime for proper handling
df["VERDATE"] = pd.to_datetime(df["VERDATE"], errors="coerce")

In [7]:
# # Check missing values per column (after initial load)
df.isna().sum().sort_values(ascending=False)

PUMFID      0
EHA_10      0
EHA_25      0
FP_05       0
DWI_05A     0
           ..
PATT_05     0
PATT_10     0
PATT_15A    0
PATT_15B    0
VERDATE     0
Length: 103, dtype: int64

In [8]:
# Get basic descriptive stats for numeric columns (unweighted, structure check)
df.describe().T

count                           mean                  min  \
PUMFID    38657.0                        82829.0              63501.0   
EHA_10    38657.0                       2.888119                  1.0   
EHA_25    38657.0                       1.953126                  1.0   
FP_05     38657.0                       1.028119                  1.0   
DWI_05A   38657.0                        1.96337                  1.0   
...           ...                            ...                  ...   
PATT_05   38657.0                       6.103733                  0.0   
PATT_10   38657.0                        1.87803                  1.0   
PATT_15A  38657.0                       5.800321                  1.0   
PATT_15B  38657.0                       5.399669                  1.0   
VERDATE     38657  2025-11-07 23:59:59.999999744  2025-11-08 00:00:00   

                          25%                  50%                  75%  \
PUMFID                73165.0              82829.0              92493.0   
EHA_10                    2.0                  3.0                  4.0   
EHA_25                    2.0                  2.0                  2.0   
FP_05                     1.0                  1.0                  1.0   
DWI_05A                   2.0                  2.0                  2.0   
...                       ...                  ...                  ...   
PATT_05                   1.0                  1.0                  2.0   
PATT_10                   1.0                  1.0                  2.0   
PATT_15A                  6.0                  6.0                  6.0   
PATT_15B                  6.0                  6.0                  6.0   
VERDATE   2025-11-08 00:00:00  2025-11-08 00:00:00  2025-11-08 00:00:00   

                          max           std  
PUMFID               102157.0  11159.459015  
EHA_10                    9.0      1.076674  
EHA_25                    9.0      0.327064  
FP_05                     9.0      0.294491  
DWI_05A                   9.0      0.446082  
...                       ...           ...  
PATT_05                  99.0     21.497314  
PATT_10                   9.0      1.892823  
PATT_15A                  9.0      1.272822  
PATT_15B                  9.0      1.733045  
VERDATE   2025-11-08 00:00:00           NaN  

[103 rows x 8 columns]

In [ ]:
# # Identify columns that contain obvious reserved codes (e.g., 9, 96, 99, 999, etc.)
reserved_codes = [9, 96, 99, 996, 999, 999.6, 999.9, 99999996, 99999999, 99999999999]

reserved_check = {
    col: df[col].isin(reserved_codes).any()
    for col in df.columns
    if df[col].dtype != "object"
}

[k for k, v in reserved_check.items() if v]


---
## [Task 1 : Handle Reserved Codes as NA](https://emmanuelolajubu90.atlassian.net/browse/SCRUM-12)

* Identify outcome vars (PCHN, PSTIR_GR) → confirm universe + eligibility rules

* Apply logic to convert reserved codes to NA for all relevant variables, without performing any recoding at this stage.

---

In [16]:
# 1. PCHN (Core Housing Need)

print("PCHN (Original Counts):")
df['PCHN'].value_counts(dropna=False).sort_index()

PCHN (Original Counts):


PCHN
1     6164
2    30938
9     1555
Name: count, dtype: int64

In [13]:
# Create a clean version: Map 9 to NaN
df['PCHN_Clean'] = df['PCHN'].replace({9: np.nan})

In [15]:
# Print the cleaned data count

print(df['PCHN_Clean'].value_counts(dropna=False).sort_index())
print('_'*50)
print(f"Records Excluded (Not Stated): {df['PCHN_Clean'].isna().sum()}")

PCHN_Clean
1.0     6164
2.0    30938
NaN     1555
Name: count, dtype: int64
__________________________________________________
Records Excluded (Not Stated): 1555


In [22]:
# 2. PSTIR_GR (Shelter-cost-to-income ratio group)

print("Definition: 1 (<30%), 2 (30-50%), 3 (50-100%), 4 (>=100%) \n")
print("Reserved Codes: 5 = Not Applicable, 9 = Not Stated \n")

print("PSTIR_GR (Original Counts):")
df['PSTIR_GR'].value_counts(dropna=False).sort_index()

Definition: 1 (<30%), 2 (30-50%), 3 (50-100%), 4 (>=100%) 

Reserved Codes: 5 = Not Applicable, 9 = Not Stated 

PSTIR_GR (Original Counts):


PSTIR_GR
1    28650
2     6440
3     2012
4      549
5      429
9      577
Name: count, dtype: int64

In [23]:
# Map 5 and 9 to NaN

df['PSTIR_GR_Clean'] = df['PSTIR_GR'].replace({5: np.nan, 9: np.nan})

In [30]:
print("\nPSTIR_GR cleaned data counts:")
display(df['PSTIR_GR_Clean'].value_counts(dropna=False).sort_index())
print(f"\nRecords Excluded (N/A or Not Stated): {df['PSTIR_GR_Clean'].isna().sum()}")


PSTIR_GR cleaned data counts:


PSTIR_GR_Clean
1.0    28650
2.0     6440
3.0     2012
4.0      549
NaN     1006
Name: count, dtype: int64


Records Excluded (N/A or Not Stated): 1006


In [33]:
# 3. Valid Universe Check : How many households are valid for BOTH measures?

valid_both = df.dropna(subset=['PCHN_Clean', 'PSTIR_GR_Clean'])
print(f"Total rows in dataset: {len(df)}")
print(f"Rows valid for BOTH PCHN and PSTIR_GR: {len(valid_both)}")

Total rows in dataset: 38657
Rows valid for BOTH PCHN and PSTIR_GR: 37102


### Output for Task 1

* **PCHN**: Preserved 37,102 valid households (6,164 In Need / 30,938 Not In Need) data. We excluded 1,555 "Not Stated" records.

* **PSTIR_GR**: Preserved 37,651 valid households. We excluded 1,006 records (429 "Not Applicable" + 577 "Not Stated") data.

* **Intersection**: 37,102 households have valid data for both variables, makes it valid sample data size for our analysis.